frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -211,219 +230,290 @@ describe("QueryBuilder", () => {
-          card: TEST_MODEL_CARD,
-          dataset: TEST_MODEL_DATASET,
-          initialRoute: `/model/${TEST_MODEL_CARD.id}/metadata`,
+          card: TEST_STRUCTURED_CARD,
         });
 
-        const columnDisplayName = await screen.findByTitle("Display name");
+        expect(screen.queryByText("Count")).not.toBeInTheDocument();
+        userEvent.click(await screen.findByText("Summarize"));
+        userEvent.click(await screen.findByText("Done"));
+        expect(await screen.findByText("Count")).toBeInTheDocument();
 
-        userEvent.click(columnDisplayName);
-        userEvent.type(columnDisplayName, "X");
+        const mockEvent = callMockEvent(mockEventListener, "beforeunload");
+        expect(mockEvent.preventDefault).not.toHaveBeenCalled();
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
+      });
 
-        await waitFor(() => {
-          expect(columnDisplayName).toHaveValue(
-            `${TEST_MODEL_DATASET_COLUMN.display_name}X`,
-          );
+      it("should not trigger beforeunload event when user tries to leave an ad-hoc native query", async () => {
+        const { mockEventListener } = await setup({
+          card: TEST_UNSAVED_STRUCTURED_CARD,
         });
 
-        userEvent.tab();
+        expect(screen.queryByText("Count")).not.toBeInTheDocument();
+        userEvent.click(await screen.findByText("Summarize"));
+        userEvent.click(await screen.findByText("Done"));
+        expect(await screen.findByText("Count")).toBeInTheDocument();
 
         const mockEvent = callMockEvent(mockEventListener, "beforeunload");
         expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-        expect(mockEvent.returnValue).toBe(undefined);
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
       });
 
-      it("should not trigger beforeunload event when model metadata is unedited", async () => {
+      it("should not trigger beforeunload event when query is unedited", async () => {
         const { mockEventListener } = await setup({
-          card: TEST_MODEL_CARD,
-          dataset: TEST_MODEL_DATASET,
-          initialRoute: `/model/${TEST_MODEL_CARD.id}/metadata`,
+          card: TEST_STRUCTURED_CARD,
         });
 
         const mockEvent = callMockEvent(mockEventListener, "beforeunload");
         expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-        expect(mockEvent.returnValue).toBe(undefined);
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
       });
     });
   });
 
-  describe("beforeunload events in native queries", () => {
-    afterEach(() => {
-      jest.restoreAllMocks();
-    });
-
-    it("should trigger beforeunload event when leaving edited question", async () => {
-      const { mockEventListener } = await setup({
+  describe("downloading results", () => {
```

Comment:
There is basically nothing change above this line, and from this line downward is the new change.